### PR TITLE
Dumping child processes

### PR DESCRIPTION
**Pull request by nohwnd**
## Description
Allow child processes to be crash and hang dumped on Windows, MacOS and Linux. Dumps are by default generated to %TEMP%\guid and all generated dumps are collected from that folder. Unlike before where we told the tool the exact path for the dump and then looked just for that file.

## Related issue
#2498

### CODE REVIEW

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessCodeMethods.cs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        }
+
+        public static List<ProcessTreeNode> GetProcessTree(this Process process)
+        {
+            var sw = Stopwatch.StartNew();
+            var childProcesses = Process.GetProcesses()
+            .Where(p => IsChildCandidate(p, process))
+            .ToList();
+
+            Console.WriteLine(sw.ElapsedMilliseconds.ToString());
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessCodeMethods.cs:52]: Removing.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessCodeMethods.cs [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+            EqtTrace.Verbose($"ProcessCodeMethods.Suspend: Suspending process {process.Id} - {process.ProcessName}.");
+            NtSuspendProcess(process.Handle);
+        }
+
+        internal static void SuspendLinuxMacOs(Process process)
+        {
+            try
+            {
+                Process.Start("kill", "-STOP {process.Id}");
+                var output = new StringBuilder();
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessCodeMethods.cs:88]: Old code, updating.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                try
+                {
+                    EqtTrace.Verbose($"NetClientHangDumper.Dump: Killing process {p.Id} - {p.ProcessName}.");
+                    p.Kill();
+                }
+                catch (Exception ex)
+                {
+                    EqtTrace.Error($"NetClientHangDumper.Dump: Error killing process {p.Id} - {p.ProcessName}: {ex}.");
+                }
+            }
```
Comment (Sanan07) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs:86]: Would not be better to place it in separate method?

Comment (nohwnd) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs:86]: Why?

Comment (Sanan07) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs:86]: For better refactoring :)

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            var processTree = process.GetProcessTree();
+
+            if (EqtTrace.IsVerboseEnabled)
+            {
+                if (processTree.Count > 1)
+                {
+                    EqtTrace.Verbose("NetClientHangDumper.Dump: Dumping this process tree (from bottom):");
+                    ConsoleOutput.Instance.Information(false, "Blame: Dumping this process tree (from bottom):");
+
+                    foreach (var p in processTree.OrderBy(t => t.Level))
```
Comment (Sanan07) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs:29]: Does that ordering happen in each iteration? Would be better to do it before the loop.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs (hunk 3 above)
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs:86]: Refactor for better refactoring?

Comment (Sanan07) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs:86]: To have better refactored code.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs (hunk 4 above)
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs:29]: It does not, it orders once, and foreach iterates over the resulting IEnumerable, if it did not you'd get collection modified exception.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs (hunk 3 above)
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/NetClientHangDumper.cs:86]: Yes, but how adding a method helps the code to be better? It invokes a single method in between logging. This imho reads okay, and the log will be easier to parse.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessCodeMethods.cs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+                IntPtr processHandle,
+                int processInformationClass,
+                out PROCESS_BASIC_INFORMATION processInformation,
+                int processInformationLength,
+                out int returnLength);
+
+        [DllImport("ntdll.dll", SetLastError = true)]
+        private static extern IntPtr NtSuspendProcess(IntPtr processHandle);
+
+        [DllImport("libpsl-native")]
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessCodeMethods.cs:267]: no, removed.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessCodeMethods.cs [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                Process ps = new Process();
+                ps.StartInfo.FileName = "ps";
+                ps.StartInfo.Arguments = $"-o ppid= {process.Id}";
+                ps.StartInfo.UseShellExecute = false;
+                ps.StartInfo.RedirectStandardOutput = true;
+                ps.OutputDataReceived += (_, e) => output.Append(e.Data);
+                ps.ErrorDataReceived += (_, e) => err.Append(e.Data);
+                ps.Start();
+                ps.BeginOutputReadLine();
+                ps.WaitForExit();
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessCodeMethods.cs:99]: added overpowered 5s timeout, just to be sure.

**Review by Haplois — APPROVED** [review]

**Pull request merged**